{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7226d652",
   "metadata": {},
   "outputs": [],
   "source": [
    "def warn(*args, **kwargs):\n",
    "    pass\n",
    "import warnings\n",
    "warnings.warn = warn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ad607155",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ce8a43b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data reading and parsing\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8b3491df",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# feature encoding\n",
    "import ifeatpro.features as ipro\n",
    "import pssmpro.features as ppro\n",
    "import ngrampro as npro\n",
    "from scipy import sparse, io\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48161eea",
   "metadata": {},
   "source": [
    "# TE raw data files read and parse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd2d5a72",
   "metadata": {},
   "source": [
    "## Raw data reading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "90037fdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_raw = \"../data/raw/TE_trainset.csv\" #rename for other datasets\n",
    "test_raw = \"../data/raw/TE_testset.csv\" # rename for other datasets\n",
    "\n",
    "\n",
    "df_train = pd.read_csv(train_raw, header=None, names=[\"enz_name\", \"enz_seq\", \"enz_label\"])\n",
    "df_test = pd.read_csv(test_raw, header=None, names=[\"enz_name\", \"enz_seq\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "20eff2ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>enz_name</th>\n",
       "      <th>enz_seq</th>\n",
       "      <th>enz_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A._hypogaea_l._(AhFatA)</td>\n",
       "      <td>MLKVSCNGSDRVQFMAQCGFAGQPASVLVRRRSVSAVGFGYPMNRV...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Arabidopsis_thaliana</td>\n",
       "      <td>MVATSATSSFFPVPSSSLDPNGKGNKIGSTNLAGLNSTPNSGRMKV...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  enz_name                                            enz_seq  \\\n",
       "0  A._hypogaea_l._(AhFatA)  MLKVSCNGSDRVQFMAQCGFAGQPASVLVRRRSVSAVGFGYPMNRV...   \n",
       "1     Arabidopsis_thaliana  MVATSATSSFFPVPSSSLDPNGKGNKIGSTNLAGLNSTPNSGRMKV...   \n",
       "\n",
       "   enz_label  \n",
       "0          1  \n",
       "1          1  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0085dff5",
   "metadata": {},
   "source": [
    "## Raw data parsing \n",
    "\n",
    "*for easier downstream applications*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0cf894c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse raw file \n",
    "# upper case all sequences\n",
    "def up_seq(seq):\n",
    "    return seq.upper().replace('-','')\n",
    "\n",
    "\n",
    "df_train[\"enz_seq\"] = df_train.enz_seq.apply(up_seq)\n",
    "df_test[\"enz_seq\"] = df_test.enz_seq.apply(up_seq)\n",
    "\n",
    "# get rid of sequences with illegitimate amino acids \n",
    "df_train = df_train.loc[~df_train[\"enz_seq\"].str.contains('B|J|O|U|X|Z')]\n",
    "df_test = df_test.loc[~df_test[\"enz_seq\"].str.contains('B|J|O|U|X|Z')]\n",
    "\n",
    "# create enzyme alias\n",
    "enz_alias_train = [f'enz_{i}' for i in range(len(df_train['enz_name']))]\n",
    "df_train = df_train.assign(enz_alias=enz_alias_train)\n",
    "enz_alias_test = [f'test_enz_{i}' for i in range(len(df_test['enz_name']))]\n",
    "df_test = df_test.assign(enz_alias=enz_alias_test)\n",
    "\n",
    "# enzyme alias to original enzyme name mapping\n",
    "enz_train_name_map = \"../data/mappings/train_enz_map.csv\"\n",
    "enz_test_name_map = \"../data/mappings/test_enz_map.csv\"\n",
    "df_train.loc[:, [\"enz_alias\", \"enz_name\"]].to_csv(enz_train_name_map, index=False, header=False)\n",
    "df_test.loc[:, [\"enz_alias\", \"enz_name\"]].to_csv(enz_test_name_map, index=False, header=False)\n",
    "\n",
    "# create fasta file of sequence\n",
    "enz_train_fasta = \"../data/seq/train_enz.fa\"\n",
    "enz_test_fasta = \"../data/seq/test_enz.fa\"\n",
    "\n",
    "train_fasta_stream = open(enz_train_fasta, \"w\")\n",
    "test_fasta_stream = open(enz_test_fasta, \"w\")\n",
    "\n",
    "for value in df_train.loc[:, [\"enz_alias\", \"enz_seq\"]].values:\n",
    "    train_fasta_stream.write(f\">{value[0]}\\n{value[1]}\\n\")\n",
    "\n",
    "for value in df_test.loc[:, [\"enz_alias\", \"enz_seq\"]].values:\n",
    "    test_fasta_stream.write(f\">{value[0]}\\n{value[1]}\\n\")\n",
    "\n",
    "train_fasta_stream.close()\n",
    "test_fasta_stream.close()\n",
    "\n",
    "# create csv file of sequence\n",
    "enz_train_csv = \"../data/seq/train_enz.csv\"\n",
    "enz_test_csv = \"../data/seq/test_enz.csv\"\n",
    "\n",
    "df_train.loc[:, [\"enz_alias\", \"enz_seq\"]].to_csv(enz_train_csv, header=False, index=False)\n",
    "df_test.loc[:, [\"enz_alias\", \"enz_seq\"]].to_csv(enz_test_csv, header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "353a02e4",
   "metadata": {},
   "source": [
    "## Creating labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "28d863ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create labels as csv\n",
    "train_labels = \"../data/label/train_enz_label.csv\"\n",
    "\n",
    "df_train.loc[:, [\"enz_alias\", \"enz_label\"]].to_csv(train_labels, index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e989feb5",
   "metadata": {},
   "source": [
    "# Numerically encoding TE sequences\n",
    "\n",
    "The TE sequences are numerically encoded in 47 different ways. They are\n",
    "\n",
    "1. 21 types of physicochemical encoding using ifeatpro tool created as a part of this project. [ifeatpro link](https://pypi.org/project/ifeatpro/)\n",
    "\n",
    "\n",
    "2. 21 types of PSSM based encodings using pssmpro tool created as a part of this project. [pssmpro link](https://pypi.org/project/pssmpro/)\n",
    "\n",
    "\n",
    "3. 2 types of ngram based encodings using ngrampro tool created as a part of this project. [ngrampro link](https://pypi.org/project/ngrampro/)\n",
    "\n",
    "\n",
    "4. 3 types of kernel based encodings using an external tool written in R known as KeBABS. [KeBABS link](https://bioconductor.org/packages/release/bioc/vignettes/kebabs/inst/doc/kebabs.pdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c69ce333",
   "metadata": {},
   "source": [
    "## ifeatpro\n",
    "\n",
    "ifeatpro [link](https://pypi.org/project/ifeatpro/) can be directly used with a fasta file that contains the protein sequences in fasta format. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "06e63d54",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function get_all_features in module ifeatpro.features:\n",
      "\n",
      "get_all_features(fasta_file, output_dir)\n",
      "    A function to create 21 numerically encoded features for protein sequences\n",
      "    :param fasta_file: The path to a file that contains all the protein sequences in fasta format\n",
      "    :param output_dir: The path to a directory where the feature encoded files will be stored\n",
      "    :return: None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(ipro.get_all_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5643ffba",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Descriptor type: aac\n",
      "Descriptor type: cksaap\n",
      "Descriptor type: tpc\n",
      "Descriptor type: dpc\n",
      "Descriptor type: dde\n",
      "Descriptor type: gaac\n",
      "Descriptor type: cksaagp\n",
      "Descriptor type: gtpc\n",
      "Descriptor type: gdpc\n",
      "Descriptor type: moran\n",
      "Descriptor type: geary\n",
      "Descriptor type: nmbroto\n",
      "Descriptor type: ctdc\n",
      "Descriptor type: ctdt\n",
      "Descriptor type: ctdd\n",
      "Descriptor type: ctriad\n",
      "Descriptor type: ksctriad\n",
      "Descriptor type: socnumber\n",
      "Descriptor type: qsorder\n",
      "Descriptor type: paac\n",
      "Descriptor type: apaac\n",
      "Descriptor type: aac\n",
      "Descriptor type: cksaap\n",
      "Descriptor type: tpc\n",
      "Descriptor type: dpc\n",
      "Descriptor type: dde\n",
      "Descriptor type: gaac\n",
      "Descriptor type: cksaagp\n",
      "Descriptor type: gtpc\n",
      "Descriptor type: gdpc\n",
      "Descriptor type: moran\n",
      "Descriptor type: geary\n",
      "Descriptor type: nmbroto\n",
      "Descriptor type: ctdc\n",
      "Descriptor type: ctdt\n",
      "Descriptor type: ctdd\n",
      "Descriptor type: ctriad\n",
      "Descriptor type: ksctriad\n",
      "Descriptor type: socnumber\n",
      "Descriptor type: qsorder\n",
      "Descriptor type: paac\n",
      "Descriptor type: apaac\n"
     ]
    }
   ],
   "source": [
    "train_fasta_file = \"../data/seq/train_enz.fa\"\n",
    "test_fasta_file = \"../data/seq/test_enz.fa\"\n",
    "\n",
    "train_output_dir = \"../data/features/ifeatpro/train/\"\n",
    "test_output_dir = \"../data/features/ifeatpro/test/\"\n",
    "\n",
    "os.makedirs(train_output_dir, exist_ok=True)\n",
    "os.makedirs(test_output_dir, exist_ok=True)\n",
    "\n",
    "ipro.features.get_all_features(train_fasta_file, train_output_dir) \n",
    "ipro.features.get_all_features(test_fasta_file, test_output_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9523fce0",
   "metadata": {},
   "source": [
    "## pssmpro\n",
    "\n",
    "pssmpro [link](https://pypi.org/project/pssmpro/) requires the pssm profile of the protein sequences as input. At first the pssm profiles need to be created and then these profiles can be numerically encoded using pssmpro provided function. pssmpro also provides a function to create the numerical encodings of proteins sequences. The psiblast program path and an indexed blast database are required as function argument."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "087c5a46",
   "metadata": {},
   "source": [
    "### Creating the pssm profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "16e9f4ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function create_pssm_profile in module pssmpro.features:\n",
      "\n",
      "create_pssm_profile(seq_file, out_dir, psiblast_exec, database_prefix, num_threads=24)\n",
      "    A function to create psiblast or pssm profile for protein sequences\n",
      "    :param seq_file: A csv file with name of the protein followed by its sequence separated by a comma\n",
      "    :param out_dir: The directory where the user would like to store the pssm profiles of all the sequences\n",
      "    :param psiblast_exec: The path of the psiblast executable. psiblast program needs to be installed\n",
      "    :param database_prefix: The path of the indexed blast database directory prefix\n",
      "    :param num_threads: Number of threads to use while creating the psiblast profile\n",
      "    :return: The output directory where the psiblast/pssm profiles are stored\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(ppro.create_pssm_profile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5007f2e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_seq_file = \"../data/seq/train_enz.csv\"\n",
    "test_seq_file = \"../data/seq/test_enz.csv\"\n",
    "psiblast_path = \"/opt/aci/sw/ncbi-rmblastn/2.9.0_gcc-8.3.1-bxy/bin/psiblast\" # provide the path to your psiblast program \n",
    "database_pre = \"../../pssmpro_test_data/uniref50/uniref50db\" # database creation described in pssmpro link given above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "addb6408",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating psiblast profile for protein: enz_0\n",
      "Generating psiblast profile for protein: enz_1\n",
      "Generating psiblast profile for protein: enz_2\n",
      "Generating psiblast profile for protein: enz_3\n",
      "Generating psiblast profile for protein: enz_4\n",
      "Generating psiblast profile for protein: enz_5\n",
      "Generating psiblast profile for protein: enz_6\n",
      "Generating psiblast profile for protein: enz_7\n",
      "Generating psiblast profile for protein: enz_8\n",
      "Generating psiblast profile for protein: enz_9\n",
      "Generating psiblast profile for protein: enz_10\n",
      "Generating psiblast profile for protein: enz_11\n",
      "Generating psiblast profile for protein: enz_12\n",
      "Generating psiblast profile for protein: enz_13\n",
      "Generating psiblast profile for protein: enz_14\n",
      "Generating psiblast profile for protein: enz_15\n",
      "Generating psiblast profile for protein: enz_16\n",
      "Generating psiblast profile for protein: enz_17\n",
      "Generating psiblast profile for protein: enz_18\n",
      "Generating psiblast profile for protein: enz_19\n",
      "Generating psiblast profile for protein: enz_20\n",
      "Generating psiblast profile for protein: enz_21\n",
      "Generating psiblast profile for protein: enz_22\n",
      "Generating psiblast profile for protein: enz_23\n",
      "Generating psiblast profile for protein: enz_24\n",
      "Generating psiblast profile for protein: enz_25\n",
      "Generating psiblast profile for protein: enz_26\n",
      "Generating psiblast profile for protein: enz_27\n",
      "Generating psiblast profile for protein: enz_28\n",
      "Generating psiblast profile for protein: enz_29\n",
      "Generating psiblast profile for protein: enz_30\n",
      "Generating psiblast profile for protein: enz_31\n",
      "Generating psiblast profile for protein: enz_32\n",
      "Generating psiblast profile for protein: enz_33\n",
      "Generating psiblast profile for protein: enz_34\n",
      "Generating psiblast profile for protein: enz_35\n",
      "Generating psiblast profile for protein: enz_36\n",
      "Generating psiblast profile for protein: enz_37\n",
      "Generating psiblast profile for protein: enz_38\n",
      "Generating psiblast profile for protein: enz_39\n",
      "Generating psiblast profile for protein: enz_40\n",
      "Generating psiblast profile for protein: enz_41\n",
      "Generating psiblast profile for protein: enz_42\n",
      "Generating psiblast profile for protein: enz_43\n",
      "Generating psiblast profile for protein: enz_44\n",
      "Generating psiblast profile for protein: enz_45\n",
      "Generating psiblast profile for protein: enz_46\n",
      "Generating psiblast profile for protein: enz_47\n",
      "Generating psiblast profile for protein: enz_48\n",
      "Generating psiblast profile for protein: enz_49\n",
      "Generating psiblast profile for protein: enz_50\n",
      "Generating psiblast profile for protein: enz_51\n",
      "Generating psiblast profile for protein: enz_52\n",
      "Generating psiblast profile for protein: enz_53\n",
      "Generating psiblast profile for protein: enz_54\n",
      "Generating psiblast profile for protein: enz_55\n",
      "Generating psiblast profile for protein: enz_56\n",
      "Generating psiblast profile for protein: enz_57\n",
      "Generating psiblast profile for protein: enz_58\n",
      "Generating psiblast profile for protein: enz_59\n",
      "Generating psiblast profile for protein: enz_60\n",
      "Generating psiblast profile for protein: enz_61\n",
      "Generating psiblast profile for protein: enz_62\n",
      "Generating psiblast profile for protein: enz_63\n",
      "Generating psiblast profile for protein: enz_64\n",
      "Generating psiblast profile for protein: enz_65\n",
      "Generating psiblast profile for protein: enz_66\n",
      "Generating psiblast profile for protein: enz_67\n",
      "Generating psiblast profile for protein: enz_68\n",
      "Generating psiblast profile for protein: enz_69\n",
      "Generating psiblast profile for protein: enz_70\n",
      "Generating psiblast profile for protein: enz_71\n",
      "Generating psiblast profile for protein: enz_72\n",
      "Generating psiblast profile for protein: enz_73\n",
      "Generating psiblast profile for protein: enz_74\n",
      "Generating psiblast profile for protein: enz_75\n",
      "Generating psiblast profile for protein: enz_76\n",
      "Generating psiblast profile for protein: enz_77\n",
      "Generating psiblast profile for protein: enz_78\n",
      "Generating psiblast profile for protein: enz_79\n",
      "Generating psiblast profile for protein: enz_80\n",
      "Generating psiblast profile for protein: enz_81\n",
      "Generating psiblast profile for protein: enz_82\n",
      "Generating psiblast profile for protein: enz_83\n",
      "Generating psiblast profile for protein: enz_84\n",
      "Generating psiblast profile for protein: enz_85\n",
      "Generating psiblast profile for protein: enz_86\n",
      "Generating psiblast profile for protein: enz_87\n",
      "Generating psiblast profile for protein: enz_88\n",
      "Generating psiblast profile for protein: enz_89\n",
      "Generating psiblast profile for protein: enz_90\n",
      "Generating psiblast profile for protein: enz_91\n",
      "Generating psiblast profile for protein: enz_92\n",
      "Generating psiblast profile for protein: enz_93\n",
      "Generating psiblast profile for protein: enz_94\n",
      "Generating psiblast profile for protein: enz_95\n",
      "Generating psiblast profile for protein: enz_96\n",
      "Generating psiblast profile for protein: enz_97\n",
      "Generating psiblast profile for protein: enz_98\n",
      "Generating psiblast profile for protein: enz_99\n",
      "Generating psiblast profile for protein: enz_100\n",
      "Generating psiblast profile for protein: enz_101\n",
      "Generating psiblast profile for protein: enz_102\n",
      "Generating psiblast profile for protein: enz_103\n",
      "Generating psiblast profile for protein: enz_104\n",
      "Generating psiblast profile for protein: enz_105\n",
      "Generating psiblast profile for protein: enz_106\n",
      "Generating psiblast profile for protein: enz_107\n",
      "Generating psiblast profile for protein: enz_108\n",
      "Generating psiblast profile for protein: enz_109\n",
      "Generating psiblast profile for protein: enz_110\n",
      "Generating psiblast profile for protein: enz_111\n",
      "Generating psiblast profile for protein: enz_112\n",
      "Generating psiblast profile for protein: enz_113\n",
      "Generating psiblast profile for protein: enz_114\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'../data/features/pssmpro/pssm_profiles'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_dir_train = \"../data/features/pssmpro/pssm_profiles/train/\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "ppro.features.create_pssm_profile(train_seq_file, output_dir_train, psiblast_path, database_pre)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ea7a7830",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating psiblast profile for protein: test_enz_0\n",
      "Generating psiblast profile for protein: test_enz_1\n",
      "Generating psiblast profile for protein: test_enz_2\n",
      "Generating psiblast profile for protein: test_enz_3\n",
      "Generating psiblast profile for protein: test_enz_4\n",
      "Generating psiblast profile for protein: test_enz_5\n",
      "Generating psiblast profile for protein: test_enz_6\n",
      "Generating psiblast profile for protein: test_enz_7\n",
      "Generating psiblast profile for protein: test_enz_8\n",
      "Generating psiblast profile for protein: test_enz_9\n",
      "Generating psiblast profile for protein: test_enz_10\n",
      "Generating psiblast profile for protein: test_enz_11\n",
      "Generating psiblast profile for protein: test_enz_12\n",
      "Generating psiblast profile for protein: test_enz_13\n",
      "Generating psiblast profile for protein: test_enz_14\n",
      "Generating psiblast profile for protein: test_enz_15\n",
      "Generating psiblast profile for protein: test_enz_16\n",
      "Generating psiblast profile for protein: test_enz_17\n",
      "Generating psiblast profile for protein: test_enz_18\n",
      "Generating psiblast profile for protein: test_enz_19\n",
      "Generating psiblast profile for protein: test_enz_20\n",
      "Generating psiblast profile for protein: test_enz_21\n",
      "Generating psiblast profile for protein: test_enz_22\n",
      "Generating psiblast profile for protein: test_enz_23\n",
      "Generating psiblast profile for protein: test_enz_24\n",
      "Generating psiblast profile for protein: test_enz_25\n",
      "Generating psiblast profile for protein: test_enz_26\n",
      "Generating psiblast profile for protein: test_enz_27\n",
      "Generating psiblast profile for protein: test_enz_28\n",
      "Generating psiblast profile for protein: test_enz_29\n",
      "Generating psiblast profile for protein: test_enz_30\n",
      "Generating psiblast profile for protein: test_enz_31\n",
      "Generating psiblast profile for protein: test_enz_32\n",
      "Generating psiblast profile for protein: test_enz_33\n",
      "Generating psiblast profile for protein: test_enz_34\n",
      "Generating psiblast profile for protein: test_enz_35\n",
      "Generating psiblast profile for protein: test_enz_36\n",
      "Generating psiblast profile for protein: test_enz_37\n",
      "Generating psiblast profile for protein: test_enz_38\n",
      "Generating psiblast profile for protein: test_enz_39\n",
      "Generating psiblast profile for protein: test_enz_40\n",
      "Generating psiblast profile for protein: test_enz_41\n",
      "Generating psiblast profile for protein: test_enz_42\n",
      "Generating psiblast profile for protein: test_enz_43\n",
      "Generating psiblast profile for protein: test_enz_44\n",
      "Generating psiblast profile for protein: test_enz_45\n",
      "Generating psiblast profile for protein: test_enz_46\n",
      "Generating psiblast profile for protein: test_enz_47\n",
      "Generating psiblast profile for protein: test_enz_48\n",
      "Generating psiblast profile for protein: test_enz_49\n",
      "Generating psiblast profile for protein: test_enz_50\n",
      "Generating psiblast profile for protein: test_enz_51\n",
      "Generating psiblast profile for protein: test_enz_52\n",
      "Generating psiblast profile for protein: test_enz_53\n",
      "Generating psiblast profile for protein: test_enz_54\n",
      "Generating psiblast profile for protein: test_enz_55\n",
      "Generating psiblast profile for protein: test_enz_56\n",
      "Generating psiblast profile for protein: test_enz_57\n",
      "Generating psiblast profile for protein: test_enz_58\n",
      "Generating psiblast profile for protein: test_enz_59\n",
      "Generating psiblast profile for protein: test_enz_60\n",
      "Generating psiblast profile for protein: test_enz_61\n",
      "Generating psiblast profile for protein: test_enz_62\n",
      "Generating psiblast profile for protein: test_enz_63\n",
      "Generating psiblast profile for protein: test_enz_64\n",
      "Generating psiblast profile for protein: test_enz_65\n",
      "Generating psiblast profile for protein: test_enz_66\n",
      "Generating psiblast profile for protein: test_enz_67\n",
      "Generating psiblast profile for protein: test_enz_68\n",
      "Generating psiblast profile for protein: test_enz_69\n",
      "Generating psiblast profile for protein: test_enz_70\n",
      "Generating psiblast profile for protein: test_enz_71\n",
      "Generating psiblast profile for protein: test_enz_72\n",
      "Generating psiblast profile for protein: test_enz_73\n",
      "Generating psiblast profile for protein: test_enz_74\n",
      "Generating psiblast profile for protein: test_enz_75\n",
      "Generating psiblast profile for protein: test_enz_76\n",
      "Generating psiblast profile for protein: test_enz_77\n",
      "Generating psiblast profile for protein: test_enz_78\n",
      "Generating psiblast profile for protein: test_enz_79\n",
      "Generating psiblast profile for protein: test_enz_80\n",
      "Generating psiblast profile for protein: test_enz_81\n",
      "Generating psiblast profile for protein: test_enz_82\n",
      "Generating psiblast profile for protein: test_enz_83\n",
      "Generating psiblast profile for protein: test_enz_84\n",
      "Generating psiblast profile for protein: test_enz_85\n",
      "Generating psiblast profile for protein: test_enz_86\n",
      "Generating psiblast profile for protein: test_enz_87\n",
      "Generating psiblast profile for protein: test_enz_88\n",
      "Generating psiblast profile for protein: test_enz_89\n",
      "Generating psiblast profile for protein: test_enz_90\n",
      "Generating psiblast profile for protein: test_enz_91\n",
      "Generating psiblast profile for protein: test_enz_92\n",
      "Generating psiblast profile for protein: test_enz_93\n",
      "Generating psiblast profile for protein: test_enz_94\n",
      "Generating psiblast profile for protein: test_enz_95\n",
      "Generating psiblast profile for protein: test_enz_96\n",
      "Generating psiblast profile for protein: test_enz_97\n",
      "Generating psiblast profile for protein: test_enz_98\n",
      "Generating psiblast profile for protein: test_enz_99\n",
      "Generating psiblast profile for protein: test_enz_100\n",
      "Generating psiblast profile for protein: test_enz_101\n",
      "Generating psiblast profile for protein: test_enz_102\n",
      "Generating psiblast profile for protein: test_enz_103\n",
      "Generating psiblast profile for protein: test_enz_104\n",
      "Generating psiblast profile for protein: test_enz_105\n",
      "Generating psiblast profile for protein: test_enz_106\n",
      "Generating psiblast profile for protein: test_enz_107\n",
      "Generating psiblast profile for protein: test_enz_108\n",
      "Generating psiblast profile for protein: test_enz_109\n",
      "Generating psiblast profile for protein: test_enz_110\n",
      "Generating psiblast profile for protein: test_enz_111\n",
      "Generating psiblast profile for protein: test_enz_112\n",
      "Generating psiblast profile for protein: test_enz_113\n",
      "Generating psiblast profile for protein: test_enz_114\n",
      "Generating psiblast profile for protein: test_enz_115\n",
      "Generating psiblast profile for protein: test_enz_116\n",
      "Generating psiblast profile for protein: test_enz_117\n",
      "Generating psiblast profile for protein: test_enz_118\n",
      "Generating psiblast profile for protein: test_enz_119\n",
      "Generating psiblast profile for protein: test_enz_120\n",
      "Generating psiblast profile for protein: test_enz_121\n",
      "Generating psiblast profile for protein: test_enz_122\n",
      "Generating psiblast profile for protein: test_enz_123\n",
      "Generating psiblast profile for protein: test_enz_124\n",
      "Generating psiblast profile for protein: test_enz_125\n",
      "Generating psiblast profile for protein: test_enz_126\n",
      "Generating psiblast profile for protein: test_enz_127\n",
      "Generating psiblast profile for protein: test_enz_128\n",
      "Generating psiblast profile for protein: test_enz_129\n",
      "Generating psiblast profile for protein: test_enz_130\n",
      "Generating psiblast profile for protein: test_enz_131\n",
      "Generating psiblast profile for protein: test_enz_132\n",
      "Generating psiblast profile for protein: test_enz_133\n",
      "Generating psiblast profile for protein: test_enz_134\n",
      "Generating psiblast profile for protein: test_enz_135\n",
      "Generating psiblast profile for protein: test_enz_136\n",
      "Generating psiblast profile for protein: test_enz_137\n",
      "Generating psiblast profile for protein: test_enz_138\n",
      "Generating psiblast profile for protein: test_enz_139\n",
      "Generating psiblast profile for protein: test_enz_140\n",
      "Generating psiblast profile for protein: test_enz_141\n",
      "Generating psiblast profile for protein: test_enz_142\n",
      "Generating psiblast profile for protein: test_enz_143\n",
      "Generating psiblast profile for protein: test_enz_144\n",
      "Generating psiblast profile for protein: test_enz_145\n",
      "Generating psiblast profile for protein: test_enz_146\n",
      "Generating psiblast profile for protein: test_enz_147\n",
      "Generating psiblast profile for protein: test_enz_148\n",
      "Generating psiblast profile for protein: test_enz_149\n",
      "Generating psiblast profile for protein: test_enz_150\n",
      "Generating psiblast profile for protein: test_enz_151\n",
      "Generating psiblast profile for protein: test_enz_152\n",
      "Generating psiblast profile for protein: test_enz_153\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating psiblast profile for protein: test_enz_154\n",
      "Generating psiblast profile for protein: test_enz_155\n",
      "Generating psiblast profile for protein: test_enz_156\n",
      "Generating psiblast profile for protein: test_enz_157\n",
      "Generating psiblast profile for protein: test_enz_158\n",
      "Generating psiblast profile for protein: test_enz_159\n",
      "Generating psiblast profile for protein: test_enz_160\n",
      "Generating psiblast profile for protein: test_enz_161\n",
      "Generating psiblast profile for protein: test_enz_162\n",
      "Generating psiblast profile for protein: test_enz_163\n",
      "Generating psiblast profile for protein: test_enz_164\n",
      "Generating psiblast profile for protein: test_enz_165\n",
      "Generating psiblast profile for protein: test_enz_166\n",
      "Generating psiblast profile for protein: test_enz_167\n",
      "Generating psiblast profile for protein: test_enz_168\n",
      "Generating psiblast profile for protein: test_enz_169\n",
      "Generating psiblast profile for protein: test_enz_170\n",
      "Generating psiblast profile for protein: test_enz_171\n",
      "Generating psiblast profile for protein: test_enz_172\n",
      "Generating psiblast profile for protein: test_enz_173\n",
      "Generating psiblast profile for protein: test_enz_174\n",
      "Generating psiblast profile for protein: test_enz_175\n",
      "Generating psiblast profile for protein: test_enz_176\n",
      "Generating psiblast profile for protein: test_enz_177\n",
      "Generating psiblast profile for protein: test_enz_178\n",
      "Generating psiblast profile for protein: test_enz_179\n",
      "Generating psiblast profile for protein: test_enz_180\n",
      "Generating psiblast profile for protein: test_enz_181\n",
      "Generating psiblast profile for protein: test_enz_182\n",
      "Generating psiblast profile for protein: test_enz_183\n",
      "Generating psiblast profile for protein: test_enz_184\n",
      "Generating psiblast profile for protein: test_enz_185\n",
      "Generating psiblast profile for protein: test_enz_186\n",
      "Generating psiblast profile for protein: test_enz_187\n",
      "Generating psiblast profile for protein: test_enz_188\n",
      "Generating psiblast profile for protein: test_enz_189\n",
      "Generating psiblast profile for protein: test_enz_190\n",
      "Generating psiblast profile for protein: test_enz_191\n",
      "Generating psiblast profile for protein: test_enz_192\n",
      "Generating psiblast profile for protein: test_enz_193\n",
      "Generating psiblast profile for protein: test_enz_194\n",
      "Generating psiblast profile for protein: test_enz_195\n",
      "Generating psiblast profile for protein: test_enz_196\n",
      "Generating psiblast profile for protein: test_enz_197\n",
      "Generating psiblast profile for protein: test_enz_198\n",
      "Generating psiblast profile for protein: test_enz_199\n",
      "Generating psiblast profile for protein: test_enz_200\n",
      "Generating psiblast profile for protein: test_enz_201\n",
      "Generating psiblast profile for protein: test_enz_202\n",
      "Generating psiblast profile for protein: test_enz_203\n",
      "Generating psiblast profile for protein: test_enz_204\n",
      "Generating psiblast profile for protein: test_enz_205\n",
      "Generating psiblast profile for protein: test_enz_206\n",
      "Generating psiblast profile for protein: test_enz_207\n",
      "Generating psiblast profile for protein: test_enz_208\n",
      "Generating psiblast profile for protein: test_enz_209\n",
      "Generating psiblast profile for protein: test_enz_210\n",
      "Generating psiblast profile for protein: test_enz_211\n",
      "Generating psiblast profile for protein: test_enz_212\n",
      "Generating psiblast profile for protein: test_enz_213\n",
      "Generating psiblast profile for protein: test_enz_214\n",
      "Generating psiblast profile for protein: test_enz_215\n",
      "Generating psiblast profile for protein: test_enz_216\n",
      "Generating psiblast profile for protein: test_enz_217\n",
      "Generating psiblast profile for protein: test_enz_218\n",
      "Generating psiblast profile for protein: test_enz_219\n",
      "Generating psiblast profile for protein: test_enz_220\n",
      "Generating psiblast profile for protein: test_enz_221\n",
      "Generating psiblast profile for protein: test_enz_222\n",
      "Generating psiblast profile for protein: test_enz_223\n",
      "Generating psiblast profile for protein: test_enz_224\n",
      "Generating psiblast profile for protein: test_enz_225\n",
      "Generating psiblast profile for protein: test_enz_226\n",
      "Generating psiblast profile for protein: test_enz_227\n",
      "Generating psiblast profile for protein: test_enz_228\n",
      "Generating psiblast profile for protein: test_enz_229\n",
      "Generating psiblast profile for protein: test_enz_230\n",
      "Generating psiblast profile for protein: test_enz_231\n",
      "Generating psiblast profile for protein: test_enz_232\n",
      "Generating psiblast profile for protein: test_enz_233\n",
      "Generating psiblast profile for protein: test_enz_234\n",
      "Generating psiblast profile for protein: test_enz_235\n",
      "Generating psiblast profile for protein: test_enz_236\n",
      "Generating psiblast profile for protein: test_enz_237\n",
      "Generating psiblast profile for protein: test_enz_238\n",
      "Generating psiblast profile for protein: test_enz_239\n",
      "Generating psiblast profile for protein: test_enz_240\n",
      "Generating psiblast profile for protein: test_enz_241\n",
      "Generating psiblast profile for protein: test_enz_242\n",
      "Generating psiblast profile for protein: test_enz_243\n",
      "Generating psiblast profile for protein: test_enz_244\n",
      "Generating psiblast profile for protein: test_enz_245\n",
      "Generating psiblast profile for protein: test_enz_246\n",
      "Generating psiblast profile for protein: test_enz_247\n",
      "Generating psiblast profile for protein: test_enz_248\n",
      "Generating psiblast profile for protein: test_enz_249\n",
      "Generating psiblast profile for protein: test_enz_250\n",
      "Generating psiblast profile for protein: test_enz_251\n",
      "Generating psiblast profile for protein: test_enz_252\n",
      "Generating psiblast profile for protein: test_enz_253\n",
      "Generating psiblast profile for protein: test_enz_254\n",
      "Generating psiblast profile for protein: test_enz_255\n",
      "Generating psiblast profile for protein: test_enz_256\n",
      "Generating psiblast profile for protein: test_enz_257\n",
      "Generating psiblast profile for protein: test_enz_258\n",
      "Generating psiblast profile for protein: test_enz_259\n",
      "Generating psiblast profile for protein: test_enz_260\n",
      "Generating psiblast profile for protein: test_enz_261\n",
      "Generating psiblast profile for protein: test_enz_262\n",
      "Generating psiblast profile for protein: test_enz_263\n",
      "Generating psiblast profile for protein: test_enz_264\n",
      "Generating psiblast profile for protein: test_enz_265\n",
      "Generating psiblast profile for protein: test_enz_266\n",
      "Generating psiblast profile for protein: test_enz_267\n",
      "Generating psiblast profile for protein: test_enz_268\n",
      "Generating psiblast profile for protein: test_enz_269\n",
      "Generating psiblast profile for protein: test_enz_270\n",
      "Generating psiblast profile for protein: test_enz_271\n",
      "Generating psiblast profile for protein: test_enz_272\n",
      "Generating psiblast profile for protein: test_enz_273\n",
      "Generating psiblast profile for protein: test_enz_274\n",
      "Generating psiblast profile for protein: test_enz_275\n",
      "Generating psiblast profile for protein: test_enz_276\n",
      "Generating psiblast profile for protein: test_enz_277\n",
      "Generating psiblast profile for protein: test_enz_278\n",
      "Generating psiblast profile for protein: test_enz_279\n",
      "Generating psiblast profile for protein: test_enz_280\n",
      "Generating psiblast profile for protein: test_enz_281\n",
      "Generating psiblast profile for protein: test_enz_282\n",
      "Generating psiblast profile for protein: test_enz_283\n",
      "Generating psiblast profile for protein: test_enz_284\n",
      "Generating psiblast profile for protein: test_enz_285\n",
      "Generating psiblast profile for protein: test_enz_286\n",
      "Generating psiblast profile for protein: test_enz_287\n",
      "Generating psiblast profile for protein: test_enz_288\n",
      "Generating psiblast profile for protein: test_enz_289\n",
      "Generating psiblast profile for protein: test_enz_290\n",
      "Generating psiblast profile for protein: test_enz_291\n",
      "Generating psiblast profile for protein: test_enz_292\n",
      "Generating psiblast profile for protein: test_enz_293\n",
      "Generating psiblast profile for protein: test_enz_294\n",
      "Generating psiblast profile for protein: test_enz_295\n",
      "Generating psiblast profile for protein: test_enz_296\n",
      "Generating psiblast profile for protein: test_enz_297\n",
      "Generating psiblast profile for protein: test_enz_298\n",
      "Generating psiblast profile for protein: test_enz_299\n",
      "Generating psiblast profile for protein: test_enz_300\n",
      "Generating psiblast profile for protein: test_enz_301\n",
      "Generating psiblast profile for protein: test_enz_302\n",
      "Generating psiblast profile for protein: test_enz_303\n",
      "Generating psiblast profile for protein: test_enz_304\n",
      "Generating psiblast profile for protein: test_enz_305\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating psiblast profile for protein: test_enz_306\n",
      "Generating psiblast profile for protein: test_enz_307\n",
      "Generating psiblast profile for protein: test_enz_308\n",
      "Generating psiblast profile for protein: test_enz_309\n",
      "Generating psiblast profile for protein: test_enz_310\n",
      "Generating psiblast profile for protein: test_enz_311\n",
      "Generating psiblast profile for protein: test_enz_312\n",
      "Generating psiblast profile for protein: test_enz_313\n",
      "Generating psiblast profile for protein: test_enz_314\n",
      "Generating psiblast profile for protein: test_enz_315\n",
      "Generating psiblast profile for protein: test_enz_316\n",
      "Generating psiblast profile for protein: test_enz_317\n",
      "Generating psiblast profile for protein: test_enz_318\n",
      "Generating psiblast profile for protein: test_enz_319\n",
      "Generating psiblast profile for protein: test_enz_320\n",
      "Generating psiblast profile for protein: test_enz_321\n",
      "Generating psiblast profile for protein: test_enz_322\n",
      "Generating psiblast profile for protein: test_enz_323\n",
      "Generating psiblast profile for protein: test_enz_324\n",
      "Generating psiblast profile for protein: test_enz_325\n",
      "Generating psiblast profile for protein: test_enz_326\n",
      "Generating psiblast profile for protein: test_enz_327\n",
      "Generating psiblast profile for protein: test_enz_328\n",
      "Generating psiblast profile for protein: test_enz_329\n",
      "Generating psiblast profile for protein: test_enz_330\n",
      "Generating psiblast profile for protein: test_enz_331\n",
      "Generating psiblast profile for protein: test_enz_332\n",
      "Generating psiblast profile for protein: test_enz_333\n",
      "Generating psiblast profile for protein: test_enz_334\n",
      "Generating psiblast profile for protein: test_enz_335\n",
      "Generating psiblast profile for protein: test_enz_336\n",
      "Generating psiblast profile for protein: test_enz_337\n",
      "Generating psiblast profile for protein: test_enz_338\n",
      "Generating psiblast profile for protein: test_enz_339\n",
      "Generating psiblast profile for protein: test_enz_340\n",
      "Generating psiblast profile for protein: test_enz_341\n",
      "Generating psiblast profile for protein: test_enz_342\n",
      "Generating psiblast profile for protein: test_enz_343\n",
      "Generating psiblast profile for protein: test_enz_344\n",
      "Generating psiblast profile for protein: test_enz_345\n",
      "Generating psiblast profile for protein: test_enz_346\n",
      "Generating psiblast profile for protein: test_enz_347\n",
      "Generating psiblast profile for protein: test_enz_348\n",
      "Generating psiblast profile for protein: test_enz_349\n",
      "Generating psiblast profile for protein: test_enz_350\n",
      "Generating psiblast profile for protein: test_enz_351\n",
      "Generating psiblast profile for protein: test_enz_352\n",
      "Generating psiblast profile for protein: test_enz_353\n",
      "Generating psiblast profile for protein: test_enz_354\n",
      "Generating psiblast profile for protein: test_enz_355\n",
      "Generating psiblast profile for protein: test_enz_356\n",
      "Generating psiblast profile for protein: test_enz_357\n",
      "Generating psiblast profile for protein: test_enz_358\n",
      "Generating psiblast profile for protein: test_enz_359\n",
      "Generating psiblast profile for protein: test_enz_360\n",
      "Generating psiblast profile for protein: test_enz_361\n",
      "Generating psiblast profile for protein: test_enz_362\n",
      "Generating psiblast profile for protein: test_enz_363\n",
      "Generating psiblast profile for protein: test_enz_364\n",
      "Generating psiblast profile for protein: test_enz_365\n",
      "Generating psiblast profile for protein: test_enz_366\n",
      "Generating psiblast profile for protein: test_enz_367\n",
      "Generating psiblast profile for protein: test_enz_368\n",
      "Generating psiblast profile for protein: test_enz_369\n",
      "Generating psiblast profile for protein: test_enz_370\n",
      "Generating psiblast profile for protein: test_enz_371\n",
      "Generating psiblast profile for protein: test_enz_372\n",
      "Generating psiblast profile for protein: test_enz_373\n",
      "Generating psiblast profile for protein: test_enz_374\n",
      "Generating psiblast profile for protein: test_enz_375\n",
      "Generating psiblast profile for protein: test_enz_376\n",
      "Generating psiblast profile for protein: test_enz_377\n",
      "Generating psiblast profile for protein: test_enz_378\n",
      "Generating psiblast profile for protein: test_enz_379\n",
      "Generating psiblast profile for protein: test_enz_380\n",
      "Generating psiblast profile for protein: test_enz_381\n",
      "Generating psiblast profile for protein: test_enz_382\n",
      "Generating psiblast profile for protein: test_enz_383\n",
      "Generating psiblast profile for protein: test_enz_384\n",
      "Generating psiblast profile for protein: test_enz_385\n",
      "Generating psiblast profile for protein: test_enz_386\n",
      "Generating psiblast profile for protein: test_enz_387\n",
      "Generating psiblast profile for protein: test_enz_388\n",
      "Generating psiblast profile for protein: test_enz_389\n",
      "Generating psiblast profile for protein: test_enz_390\n",
      "Generating psiblast profile for protein: test_enz_391\n",
      "Generating psiblast profile for protein: test_enz_392\n",
      "Generating psiblast profile for protein: test_enz_393\n",
      "Generating psiblast profile for protein: test_enz_394\n",
      "Generating psiblast profile for protein: test_enz_395\n",
      "Generating psiblast profile for protein: test_enz_396\n",
      "Generating psiblast profile for protein: test_enz_397\n",
      "Generating psiblast profile for protein: test_enz_398\n",
      "Generating psiblast profile for protein: test_enz_399\n",
      "Generating psiblast profile for protein: test_enz_400\n",
      "Generating psiblast profile for protein: test_enz_401\n",
      "Generating psiblast profile for protein: test_enz_402\n",
      "Generating psiblast profile for protein: test_enz_403\n",
      "Generating psiblast profile for protein: test_enz_404\n",
      "Generating psiblast profile for protein: test_enz_405\n",
      "Generating psiblast profile for protein: test_enz_406\n",
      "Generating psiblast profile for protein: test_enz_407\n",
      "Generating psiblast profile for protein: test_enz_408\n",
      "Generating psiblast profile for protein: test_enz_409\n",
      "Generating psiblast profile for protein: test_enz_410\n",
      "Generating psiblast profile for protein: test_enz_411\n",
      "Generating psiblast profile for protein: test_enz_412\n",
      "Generating psiblast profile for protein: test_enz_413\n",
      "Generating psiblast profile for protein: test_enz_414\n",
      "Generating psiblast profile for protein: test_enz_415\n",
      "Generating psiblast profile for protein: test_enz_416\n",
      "Generating psiblast profile for protein: test_enz_417\n",
      "Generating psiblast profile for protein: test_enz_418\n",
      "Generating psiblast profile for protein: test_enz_419\n",
      "Generating psiblast profile for protein: test_enz_420\n",
      "Generating psiblast profile for protein: test_enz_421\n",
      "Generating psiblast profile for protein: test_enz_422\n",
      "Generating psiblast profile for protein: test_enz_423\n",
      "Generating psiblast profile for protein: test_enz_424\n",
      "Generating psiblast profile for protein: test_enz_425\n",
      "Generating psiblast profile for protein: test_enz_426\n",
      "Generating psiblast profile for protein: test_enz_427\n",
      "Generating psiblast profile for protein: test_enz_428\n",
      "Generating psiblast profile for protein: test_enz_429\n",
      "Generating psiblast profile for protein: test_enz_430\n",
      "Generating psiblast profile for protein: test_enz_431\n",
      "Generating psiblast profile for protein: test_enz_432\n",
      "Generating psiblast profile for protein: test_enz_433\n",
      "Generating psiblast profile for protein: test_enz_434\n",
      "Generating psiblast profile for protein: test_enz_435\n",
      "Generating psiblast profile for protein: test_enz_436\n",
      "Generating psiblast profile for protein: test_enz_437\n",
      "Generating psiblast profile for protein: test_enz_438\n",
      "Generating psiblast profile for protein: test_enz_439\n",
      "Generating psiblast profile for protein: test_enz_440\n",
      "Generating psiblast profile for protein: test_enz_441\n",
      "Generating psiblast profile for protein: test_enz_442\n",
      "Generating psiblast profile for protein: test_enz_443\n",
      "Generating psiblast profile for protein: test_enz_444\n",
      "Generating psiblast profile for protein: test_enz_445\n",
      "Generating psiblast profile for protein: test_enz_446\n",
      "Generating psiblast profile for protein: test_enz_447\n",
      "Generating psiblast profile for protein: test_enz_448\n",
      "Generating psiblast profile for protein: test_enz_449\n",
      "Generating psiblast profile for protein: test_enz_450\n",
      "Generating psiblast profile for protein: test_enz_451\n",
      "Generating psiblast profile for protein: test_enz_452\n",
      "Generating psiblast profile for protein: test_enz_453\n",
      "Generating psiblast profile for protein: test_enz_454\n",
      "Generating psiblast profile for protein: test_enz_455\n",
      "Generating psiblast profile for protein: test_enz_456\n",
      "Generating psiblast profile for protein: test_enz_457\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating psiblast profile for protein: test_enz_458\n",
      "Generating psiblast profile for protein: test_enz_459\n",
      "Generating psiblast profile for protein: test_enz_460\n",
      "Generating psiblast profile for protein: test_enz_461\n",
      "Generating psiblast profile for protein: test_enz_462\n",
      "Generating psiblast profile for protein: test_enz_463\n",
      "Generating psiblast profile for protein: test_enz_464\n",
      "Generating psiblast profile for protein: test_enz_465\n",
      "Generating psiblast profile for protein: test_enz_466\n",
      "Generating psiblast profile for protein: test_enz_467\n",
      "Generating psiblast profile for protein: test_enz_468\n",
      "Generating psiblast profile for protein: test_enz_469\n",
      "Generating psiblast profile for protein: test_enz_470\n",
      "Generating psiblast profile for protein: test_enz_471\n",
      "Generating psiblast profile for protein: test_enz_472\n",
      "Generating psiblast profile for protein: test_enz_473\n",
      "Generating psiblast profile for protein: test_enz_474\n",
      "Generating psiblast profile for protein: test_enz_475\n",
      "Generating psiblast profile for protein: test_enz_476\n",
      "Generating psiblast profile for protein: test_enz_477\n",
      "Generating psiblast profile for protein: test_enz_478\n",
      "Generating psiblast profile for protein: test_enz_479\n",
      "Generating psiblast profile for protein: test_enz_480\n",
      "Generating psiblast profile for protein: test_enz_481\n",
      "Generating psiblast profile for protein: test_enz_482\n",
      "Generating psiblast profile for protein: test_enz_483\n",
      "Generating psiblast profile for protein: test_enz_484\n",
      "Generating psiblast profile for protein: test_enz_485\n",
      "Generating psiblast profile for protein: test_enz_486\n",
      "Generating psiblast profile for protein: test_enz_487\n",
      "Generating psiblast profile for protein: test_enz_488\n",
      "Generating psiblast profile for protein: test_enz_489\n",
      "Generating psiblast profile for protein: test_enz_490\n",
      "Generating psiblast profile for protein: test_enz_491\n",
      "Generating psiblast profile for protein: test_enz_492\n",
      "Generating psiblast profile for protein: test_enz_493\n",
      "Generating psiblast profile for protein: test_enz_494\n",
      "Generating psiblast profile for protein: test_enz_495\n",
      "Generating psiblast profile for protein: test_enz_496\n",
      "Generating psiblast profile for protein: test_enz_497\n",
      "Generating psiblast profile for protein: test_enz_498\n",
      "Generating psiblast profile for protein: test_enz_499\n",
      "Generating psiblast profile for protein: test_enz_500\n",
      "Generating psiblast profile for protein: test_enz_501\n",
      "Generating psiblast profile for protein: test_enz_502\n",
      "Generating psiblast profile for protein: test_enz_503\n",
      "Generating psiblast profile for protein: test_enz_504\n",
      "Generating psiblast profile for protein: test_enz_505\n",
      "Generating psiblast profile for protein: test_enz_506\n",
      "Generating psiblast profile for protein: test_enz_507\n",
      "Generating psiblast profile for protein: test_enz_508\n",
      "Generating psiblast profile for protein: test_enz_509\n",
      "Generating psiblast profile for protein: test_enz_510\n",
      "Generating psiblast profile for protein: test_enz_511\n",
      "Generating psiblast profile for protein: test_enz_512\n",
      "Generating psiblast profile for protein: test_enz_513\n",
      "Generating psiblast profile for protein: test_enz_514\n",
      "Generating psiblast profile for protein: test_enz_515\n",
      "Generating psiblast profile for protein: test_enz_516\n",
      "Generating psiblast profile for protein: test_enz_517\n",
      "Generating psiblast profile for protein: test_enz_518\n",
      "Generating psiblast profile for protein: test_enz_519\n",
      "Generating psiblast profile for protein: test_enz_520\n",
      "Generating psiblast profile for protein: test_enz_521\n",
      "Generating psiblast profile for protein: test_enz_522\n",
      "Generating psiblast profile for protein: test_enz_523\n",
      "Generating psiblast profile for protein: test_enz_524\n",
      "Generating psiblast profile for protein: test_enz_525\n",
      "Generating psiblast profile for protein: test_enz_526\n",
      "Generating psiblast profile for protein: test_enz_527\n",
      "Generating psiblast profile for protein: test_enz_528\n",
      "Generating psiblast profile for protein: test_enz_529\n",
      "Generating psiblast profile for protein: test_enz_530\n",
      "Generating psiblast profile for protein: test_enz_531\n",
      "Generating psiblast profile for protein: test_enz_532\n",
      "Generating psiblast profile for protein: test_enz_533\n",
      "Generating psiblast profile for protein: test_enz_534\n",
      "Generating psiblast profile for protein: test_enz_535\n",
      "Generating psiblast profile for protein: test_enz_536\n",
      "Generating psiblast profile for protein: test_enz_537\n",
      "Generating psiblast profile for protein: test_enz_538\n",
      "Generating psiblast profile for protein: test_enz_539\n",
      "Generating psiblast profile for protein: test_enz_540\n",
      "Generating psiblast profile for protein: test_enz_541\n",
      "Generating psiblast profile for protein: test_enz_542\n",
      "Generating psiblast profile for protein: test_enz_543\n",
      "Generating psiblast profile for protein: test_enz_544\n",
      "Generating psiblast profile for protein: test_enz_545\n",
      "Generating psiblast profile for protein: test_enz_546\n",
      "Generating psiblast profile for protein: test_enz_547\n",
      "Generating psiblast profile for protein: test_enz_548\n",
      "Generating psiblast profile for protein: test_enz_549\n",
      "Generating psiblast profile for protein: test_enz_550\n",
      "Generating psiblast profile for protein: test_enz_551\n",
      "Generating psiblast profile for protein: test_enz_552\n",
      "Generating psiblast profile for protein: test_enz_553\n",
      "Generating psiblast profile for protein: test_enz_554\n",
      "Generating psiblast profile for protein: test_enz_555\n",
      "Generating psiblast profile for protein: test_enz_556\n",
      "Generating psiblast profile for protein: test_enz_557\n",
      "Generating psiblast profile for protein: test_enz_558\n",
      "Generating psiblast profile for protein: test_enz_559\n",
      "Generating psiblast profile for protein: test_enz_560\n",
      "Generating psiblast profile for protein: test_enz_561\n",
      "Generating psiblast profile for protein: test_enz_562\n",
      "Generating psiblast profile for protein: test_enz_563\n",
      "Generating psiblast profile for protein: test_enz_564\n",
      "Generating psiblast profile for protein: test_enz_565\n",
      "Generating psiblast profile for protein: test_enz_566\n",
      "Generating psiblast profile for protein: test_enz_567\n",
      "Generating psiblast profile for protein: test_enz_568\n",
      "Generating psiblast profile for protein: test_enz_569\n",
      "Generating psiblast profile for protein: test_enz_570\n",
      "Generating psiblast profile for protein: test_enz_571\n",
      "Generating psiblast profile for protein: test_enz_572\n",
      "Generating psiblast profile for protein: test_enz_573\n",
      "Generating psiblast profile for protein: test_enz_574\n",
      "Generating psiblast profile for protein: test_enz_575\n",
      "Generating psiblast profile for protein: test_enz_576\n",
      "Generating psiblast profile for protein: test_enz_577\n",
      "Generating psiblast profile for protein: test_enz_578\n",
      "Generating psiblast profile for protein: test_enz_579\n",
      "Generating psiblast profile for protein: test_enz_580\n",
      "Generating psiblast profile for protein: test_enz_581\n",
      "Generating psiblast profile for protein: test_enz_582\n",
      "Generating psiblast profile for protein: test_enz_583\n",
      "Generating psiblast profile for protein: test_enz_584\n",
      "Generating psiblast profile for protein: test_enz_585\n",
      "Generating psiblast profile for protein: test_enz_586\n",
      "Generating psiblast profile for protein: test_enz_587\n",
      "Generating psiblast profile for protein: test_enz_588\n",
      "Generating psiblast profile for protein: test_enz_589\n",
      "Generating psiblast profile for protein: test_enz_590\n",
      "Generating psiblast profile for protein: test_enz_591\n",
      "Generating psiblast profile for protein: test_enz_592\n",
      "Generating psiblast profile for protein: test_enz_593\n",
      "Generating psiblast profile for protein: test_enz_594\n",
      "Generating psiblast profile for protein: test_enz_595\n",
      "Generating psiblast profile for protein: test_enz_596\n",
      "Generating psiblast profile for protein: test_enz_597\n",
      "Generating psiblast profile for protein: test_enz_598\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'../data/features/pssmpro/pssm_profiles/test/'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_dir_test = \"../data/features/pssmpro/pssm_profiles/test/\"\n",
    "os.makedirs(output_dir_test, exist_ok=True)\n",
    "ppro.create_pssm_profile(test_seq_file, output_dir_test, psiblast_path, database_pre)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc8bfd3c",
   "metadata": {},
   "source": [
    "### Generating features from the profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c9448c3e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function get_all_features in module pssmpro.features:\n",
      "\n",
      "get_all_features(pssm_dir, store_dir='./')\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(ppro.get_all_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2d662db2",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_out_dir = \"../data/features/pssmpro/train/\"\n",
    "test_out_dir = \"../data/features/pssmpro/test/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6bb15e97",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(train_out_dir, exist_ok=True)\n",
    "ppro.get_all_features(\"../data/features/pssmpro/pssm_profiles/train/\", train_out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9f3bcc34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['aac_pssm',\n",
       " 'aadp_pssm',\n",
       " 'aatp',\n",
       " 'ab_pssm',\n",
       " 'd_fpssm',\n",
       " 'dp_pssm',\n",
       " 'dpc_pssm',\n",
       " 'edp',\n",
       " 'eedp',\n",
       " 'k_separated_bigrams_pssm',\n",
       " 'medp',\n",
       " 'pse_pssm',\n",
       " 'pssm_ac',\n",
       " 'pssm_cc',\n",
       " 'pssm_composition',\n",
       " 'rpm_pssm',\n",
       " 'rpssm',\n",
       " 's_fpssm',\n",
       " 'smoothed_pssm',\n",
       " 'tpc',\n",
       " 'tri_gram_pssm']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.makedirs(test_out_dir, exist_ok=True)\n",
    "ppro.get_all_features(\"../data/features/pssmpro/pssm_profiles/test/\", test_out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddd1c801",
   "metadata": {},
   "source": [
    "## ngrampro\n",
    "\n",
    "ngrampro [link](https://pypi.org/project/ngrampro/) based features are created online during model training. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5c03ef9",
   "metadata": {},
   "source": [
    "## KeBABS\n",
    "\n",
    "Kernel based features are created using an external software package called KeBABS. The Rscript used to generate the features called *Kernels-KeBABs.r* is given in the *utils* directory. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4954b26",
   "metadata": {},
   "source": [
    "### Running Kebabs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "364361bd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading required package: Biostrings\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,\n",
      "    tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: XVector\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "Loading required package: kernlab\n",
      "\n",
      "Attaching package: 'kernlab'\n",
      "\n",
      "The following object is masked from 'package:Biostrings':\n",
      "\n",
      "    type\n",
      "\n",
      "The following object is masked from 'package:BiocGenerics':\n",
      "\n",
      "    type\n",
      "\n",
      "Warning message:\n",
      "In .recacheSubclasses(def@className, def, env) :\n",
      "  undefined subclass \"numericVector\" of class \"Mnumeric\"; definition not updated\n",
      "\n",
      "Attaching package: 'Matrix'\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    expand\n",
      "\n",
      "Warning message:\n",
      "In .Call(\"freeHeapCallocsC\", as.integer(kernelType[class(kernel)])) :\n",
      "  converting NULL pointer to R NULL\n",
      "NULL\n",
      "NULL\n",
      "Warning message:\n",
      "In .Call(\"freeHeapCallocsC\", as.integer(kernelType[class(kernel)])) :\n",
      "  converting NULL pointer to R NULL\n",
      "NULL\n"
     ]
    }
   ],
   "source": [
    "# Running the Rscript using a bash command\n",
    "\n",
    "!Rscript ../utils/Kernels-KeBABs.r"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3ee678f",
   "metadata": {},
   "source": [
    "### Parsing Kebabs output\n",
    "\n",
    "*required for downstream analysis* "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a704e79a",
   "metadata": {},
   "outputs": [],
   "source": [
    "featfile_dir = \"../data/features/kernel/\"\n",
    "file_prefixes = ['spec', 'gap', 'mism']\n",
    "\n",
    "output_train_file = \"../data/features/kernel/train/\"\n",
    "output_test_file = \"../data/features/kernel/test/\"\n",
    "\n",
    "os.makedirs(output_train_file, exist_ok=True)\n",
    "os.makedirs(output_test_file, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "64c2ccfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_feat_vec_files(file_dir, outdir_train, outdir_test, file_prefix):\n",
    "    sp_mat_file = file_dir + file_prefix + \"/\" + file_prefix + '_kern_sparsematrix.txt'\n",
    "    enz_name_file = file_dir + file_prefix + \"/\" + file_prefix + '_kern_rownames.txt'\n",
    "    \n",
    "    sp_mat = io.mmread(sp_mat_file).tocsr()\n",
    "    enz_names = np.genfromtxt(enz_name_file, dtype=str)\n",
    "    \n",
    "    \n",
    "    train_enz_idx = []\n",
    "    test_enz_idx = []\n",
    "\n",
    "    for idx, enz_name in enumerate(enz_names):\n",
    "        if enz_name.startswith('enz'):\n",
    "            train_enz_idx.append(idx)\n",
    "        elif enz_name.startswith('test'):\n",
    "            test_enz_idx.append(idx)\n",
    "        else:\n",
    "            raise ValueError('Wrong Enzyme Prefix')\n",
    "            \n",
    "    X_train, X_test = sp_mat[train_enz_idx,:], sp_mat[test_enz_idx,:]\n",
    "\n",
    "    enz_names_train, enz_names_test = enz_names[train_enz_idx], enz_names[test_enz_idx]\n",
    "    \n",
    "    assert X_train.shape[0] == len(enz_names_train)\n",
    "    assert X_test.shape[0] == len(enz_names_test)\n",
    "\n",
    "    \n",
    "    sparse.save_npz(outdir_train+file_prefix+'mat.npz', X_train)\n",
    "    sparse.save_npz(outdir_test+file_prefix+'mat.npz', X_test)\n",
    "\n",
    "    np.savetxt(outdir_train+file_prefix+'enz_names.txt', enz_names_train, fmt='%s')\n",
    "    np.savetxt(outdir_test+file_prefix+'enz_names.txt', enz_names_test, fmt='%s')\n",
    "\n",
    "    \n",
    "    return "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b067b68a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for fp in file_prefixes:\n",
    "    save_feat_vec_files(featfile_dir, output_train_file, \n",
    "                        output_test_file, fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c36815ad",
   "metadata": {},
   "source": [
    "# EnZymClass\n",
    "\n",
    "**En**semble model for en**Zym**e **Class**ification\n",
    "\n",
    "## Training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aaebe8dc",
   "metadata": {},
   "source": [
    "### Best base learner decider"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4c4fea23",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9abbff3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import helper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1174e536",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import numpy as np\n",
    "import multiprocessing as mp\n",
    "from itertools import groupby\n",
    "from collections import Counter\n",
    "import pandas as pd\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "63f82932",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_enz_seq_file = \"../data/seq/train_enz.csv\"\n",
    "test_enz_seq_file = \"../data/seq/test_enz.csv\"\n",
    "label_file = \"../data/label/train_enz_label.csv\"\n",
    "train_feature_dirs = [\"../data/features/ifeatpro/train/\", \n",
    "                      \"../data/features/kernel/train/\",\n",
    "                      \"../data/features/pssmpro/train/\"]\n",
    "test_feature_dirs = [\"../data/features/ifeatpro/test/\",\n",
    "                    \"../data/features/kernel/test/\",\n",
    "                    \"../data/features/pssmpro/test/\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3eb4869f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Num_sim = 10000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b5c0fe99",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_performance(enz_file, test_enz_file, label_file, train_feat_dirs, test_feat_dirs, hyper_param_file, base_algo, k, opt, N):\n",
    "    pool = mp.Pool(mp.cpu_count())\n",
    "    \n",
    "    iter_svm = zip([enz_file for _ in range(N)],\n",
    "                   [test_enz_file for _ in range(N)],\n",
    "                   [label_file for _ in range(N)],\n",
    "                   [train_feat_dirs for _ in range(N)],\n",
    "                   [test_feat_dirs for _ in range(N)],\n",
    "                   [hyper_param_file for _ in range(N)],\n",
    "                   [base_algo for _ in range(N)],\n",
    "                   [k for _ in range(N)],\n",
    "                   [opt for _ in range(N)],\n",
    "                   range(N))\n",
    "    \n",
    "    metrics = pool.starmap(helper.check_base_performance,iter_svm)\n",
    "    \n",
    "    precision = [m[0] for m in metrics]\n",
    "    recall = [m[1] for m in metrics]\n",
    "    accuracy = [m[2] for m in metrics]\n",
    "    \n",
    "    return round(np.mean(precision), 2), round(np.mean(recall), 2), round(np.mean(accuracy), 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0a2d9224",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Checking the performance of SVM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d46da08a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 716 ms, sys: 218 ms, total: 934 ms\n",
      "Wall time: 1h 51min 26s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(0.88, 0.89, 0.8)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "check_performance(train_enz_seq_file, None, label_file, train_feature_dirs, None, None, 'SVM', 5, False, Num_sim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2377e3cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Checking the performance of NN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "267ae851",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 646 ms, sys: 242 ms, total: 888 ms\n",
      "Wall time: 2h 4min 47s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(0.81, 0.94, 0.79)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "check_performance(train_enz_seq_file, None, label_file, train_feature_dirs, None, None, 'NN', 5, False, Num_sim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "df597152",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Checking the performance of GBC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ac379c0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 504 ms, sys: 225 ms, total: 729 ms\n",
      "Wall time: 1h 55min 56s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(0.82, 0.92, 0.79)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "check_performance(train_enz_seq_file, None, label_file, train_feature_dirs, None, None, 'GBC', 5, False, Num_sim)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a3991da",
   "metadata": {},
   "source": [
    "### Individual Hyperparameter Optimization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a81726b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def most_frequent(arr):\n",
    "    count = Counter(arr)\n",
    "    most_freq = count.most_common(1)[0][0]\n",
    "    return most_freq\n",
    "\n",
    "\n",
    "def store_best_hps(enz_file, test_enz_file, label_file, train_feat_dirs, test_feat_dirs, hyper_param_file, base_algo, k, opt, N):\n",
    "    pool = mp.Pool(mp.cpu_count())\n",
    "    \n",
    "    iter_func = zip([enz_file for _ in range(N)],\n",
    "               [test_enz_file for _ in range(N)],\n",
    "               [label_file for _ in range(N)],\n",
    "               [train_feat_dirs for _ in range(N)],\n",
    "               [test_feat_dirs for _ in range(N)],\n",
    "               [hyper_param_file for _ in range(N)],\n",
    "               [base_algo for _ in range(N)],\n",
    "               [k for _ in range(N)],\n",
    "               [opt for _ in range(N)],\n",
    "               range(N))\n",
    "    \n",
    "    all_hps = sum(list(pool.starmap(helper.best_hps, iter_func)),[])\n",
    "    all_hps = sorted(all_hps, key=lambda x: x[0])\n",
    "\n",
    "    with open('../data/results/hpopt/IndHPOpt.csv','w') as f:\n",
    "        f.write('feat_name,regC,kernel,pca_comp')\n",
    "        f.write('\\n')\n",
    "        for feat_name, feat_info in groupby(all_hps, key=lambda x: x[0]):\n",
    "            best_hp_list = [hp[1] for hp in list(feat_info)]\n",
    "            best_regC = most_frequent([x[0] for x in best_hp_list])\n",
    "            best_kernel = most_frequent([x[1] for x in best_hp_list])\n",
    "            best_ncomp = most_frequent([x[2] for x in best_hp_list])\n",
    "            f.write(f\"{feat_name},{best_regC},{best_kernel},{best_ncomp}\")\n",
    "            f.write('\\n')\n",
    "\n",
    "    return "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "39b73ecd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.5 s, sys: 963 ms, total: 3.46 s\n",
      "Wall time: 11h 59min 14s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "store_best_hps(train_enz_seq_file, None, label_file, train_feature_dirs, None, None, 'SVM', 5, True, Num_sim//10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "166687ec",
   "metadata": {},
   "source": [
    "### Feature extraction technique performance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "235221dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "indhpoptfile = '../data/results/hpopt/IndHPOpt.csv'\n",
    "\n",
    "\n",
    "def indfeat_measure(enz_file, test_enz_file, label_file, train_feat_dirs, test_feat_dirs, hyper_param_file, base_algo, k, opt, N):\n",
    "    pool = mp.Pool(mp.cpu_count())\n",
    "    \n",
    "    iter_func = zip([enz_file for _ in range(N)],\n",
    "               [test_enz_file for _ in range(N)],\n",
    "               [label_file for _ in range(N)],\n",
    "               [train_feat_dirs for _ in range(N)],\n",
    "               [test_feat_dirs for _ in range(N)],\n",
    "               [hyper_param_file for _ in range(N)],\n",
    "               [base_algo for _ in range(N)],\n",
    "               [k for _ in range(N)],\n",
    "               [opt for _ in range(N)],\n",
    "               range(N))\n",
    "    \n",
    "    all_metrics = sum(list(pool.starmap(helper.indfeat_performance, iter_func)), [])\n",
    "    all_metrics = sorted(all_metrics, key=lambda x: x[0])\n",
    "    with open(\"../data/results/indfeatreport.csv\", 'w') as f:\n",
    "        f.write('featname,min_precision,max_precision,mean_precision,std_precision,min_recall,max_recall,mean_recall,std_recall,min_accuracy,max_accuracy,mean_accuracy,std_accuracy')\n",
    "        f.write('\\n')\n",
    "        for featname, featinfo in groupby(all_metrics, key=lambda x:x[0]):\n",
    "            all_metric_list = [met[1] for met in featinfo]\n",
    "            all_prec = [m[0] for m in all_metric_list]\n",
    "            all_rec = [m[1] for m in all_metric_list]\n",
    "            all_acc = [m[2] for m in all_metric_list]\n",
    "\n",
    "            min_prec = round(min(all_prec),2)\n",
    "            max_prec = round(max(all_prec),2)\n",
    "            mean_prec = round(np.mean(all_prec),2)\n",
    "            std_prec = round(np.std(all_prec),2)\n",
    "\n",
    "            min_rec = round(min(all_rec),2)\n",
    "            max_rec = round(max(all_rec),2)\n",
    "            mean_rec = round(np.mean(all_rec),2)\n",
    "            std_rec = round(np.std(all_rec),2)\n",
    "\n",
    "            min_acc = round(min(all_acc),2)\n",
    "            max_acc = round(max(all_acc),2)\n",
    "            mean_acc = round(np.mean(all_acc),2)\n",
    "            std_acc = round(np.std(all_acc),2)\n",
    "            \n",
    "            f.write(f'{featname},{min_prec},{max_prec},{mean_prec},{std_prec},{min_rec},{max_rec},{mean_rec},{std_rec},{min_acc},{max_acc},{mean_acc},{std_acc}')\n",
    "            f.write('\\n')\n",
    "\n",
    "        \n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ececde4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 34.2 s, sys: 3.81 s, total: 38 s\n",
      "Wall time: 1h 47min 25s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "indfeat_measure(train_enz_seq_file, None, label_file, train_feature_dirs, None, indhpoptfile, 'SVM', 5, False, Num_sim)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8818f115",
   "metadata": {},
   "source": [
    "### Parametric sweep of ensemble model hyperparameter k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a06662f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ensemble_param_sweep(enz_file, test_enz_file, label_file, train_feat_dirs, test_feat_dirs, hyper_param_file, base_algo, ks, opt, N):\n",
    "\n",
    "    _iter = list(itertools.product(ks, range(N)))\n",
    "    iter_func = zip([enz_file for _ in range(len(_iter))],\n",
    "           [test_enz_file for _ in range(len(_iter))],\n",
    "           [label_file for _ in range(len(_iter))],\n",
    "           [train_feat_dirs for _ in range(len(_iter))],\n",
    "           [test_feat_dirs for _ in range(len(_iter))],\n",
    "           [hyper_param_file for _ in range(len(_iter))],\n",
    "           [base_algo for _ in range(len(_iter))],\n",
    "           [k[0] for k in _iter],\n",
    "           [opt for _ in _iter],\n",
    "           [rs[1] for rs in _iter])\n",
    "    \n",
    "    pool = mp.Pool(mp.cpu_count())\n",
    "    all_preds = list(pool.starmap(helper.ensemble_pred, iter_func))\n",
    "    all_metrics = list(map(helper.get_metrics, all_preds))\n",
    "    def get_mean(met):\n",
    "        return round(np.mean(met), 2)\n",
    "    \n",
    "    with open('../data/results/en_param_sweep.csv','w') as f:\n",
    "        f.write('model_k,mean_precision,mean_recall,mean_accuracy')\n",
    "        f.write('\\n')\n",
    "        for model_k, model_info in itertools.groupby(zip(list(_iter), all_metrics), key=lambda x:x[0][0]):\n",
    "            model_metrics = [m[1] for m in list(model_info)]\n",
    "            prec = [m[0] for m in model_metrics]\n",
    "            rec = [m[1] for m in model_metrics]\n",
    "            acc = [m[2] for m in model_metrics]    \n",
    "        \n",
    "            f.write(f\"{model_k},{','.join(list(map(str,(list(map(get_mean,[prec,rec,acc]))))))}\")\n",
    "            f.write('\\n')\n",
    "        \n",
    "    return "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8cc13897",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 19s, sys: 591 ms, total: 1min 20s\n",
      "Wall time: 8h 53min 48s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "ensemble_param_sweep(train_enz_seq_file, None, label_file, train_feature_dirs, None, indhpoptfile, 'SVM', [5,9,15,21,31], False, Num_sim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "55fe2925",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model_k</th>\n",
       "      <th>mean_precision</th>\n",
       "      <th>mean_recall</th>\n",
       "      <th>mean_accuracy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>31</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.78</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   model_k  mean_precision  mean_recall  mean_accuracy\n",
       "0        5            0.87         0.89           0.80\n",
       "1        9            0.87         0.88           0.79\n",
       "2       15            0.87         0.87           0.79\n",
       "3       21            0.86         0.87           0.78\n",
       "4       31            0.85         0.87           0.78"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('../data/results/en_param_sweep.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5477193",
   "metadata": {},
   "source": [
    "### Ensemble performance evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b4f132d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ensemble_eval(enz_file, test_enz_file, label_file, train_feat_dirs, test_feat_dirs, hyper_param_file, base_algo, k, opt, N):\n",
    "\n",
    "    pool = mp.Pool(mp.cpu_count())\n",
    "    \n",
    "    iter_func = zip([enz_file for _ in range(N)],\n",
    "               [test_enz_file for _ in range(N)],\n",
    "               [label_file for _ in range(N)],\n",
    "               [train_feat_dirs for _ in range(N)],\n",
    "               [test_feat_dirs for _ in range(N)],\n",
    "               [hyper_param_file for _ in range(N)],\n",
    "               [base_algo for _ in range(N)],\n",
    "               [k for _ in range(N)],\n",
    "               [opt for _ in range(N)],\n",
    "               range(N))\n",
    "    \n",
    "    all_preds = list(pool.starmap(helper.ensemble_pred, iter_func))\n",
    "    model_metrics = list(map(helper.get_metrics, all_preds))\n",
    "    all_prec = [m[0] for m in model_metrics]\n",
    "    all_rec = [m[1] for m in model_metrics]\n",
    "    all_acc = [m[2] for m in model_metrics]\n",
    "    \n",
    "    min_prec = round(min(all_prec),2)\n",
    "    max_prec = round(max(all_prec),2)\n",
    "    mean_prec = round(np.mean(all_prec),2)\n",
    "    std_prec = round(np.std(all_prec),2)\n",
    "\n",
    "    min_rec = round(min(all_rec),2)\n",
    "    max_rec = round(max(all_rec),2)\n",
    "    mean_rec = round(np.mean(all_rec),2)\n",
    "    std_rec = round(np.std(all_rec),2)\n",
    "\n",
    "    min_acc = round(min(all_acc),2)\n",
    "    max_acc = round(max(all_acc),2)\n",
    "    mean_acc = round(np.mean(all_acc),2)\n",
    "    std_acc = round(np.std(all_acc),2)\n",
    "    \n",
    "    with open('../data/results/ensemble_results.csv','w') as f:\n",
    "        for prec, rec, acc in model_metrics:\n",
    "            f.write(f\"{prec},{rec},{acc}\\n\")\n",
    "    \n",
    "    with open('../data/results/ensemble_report.csv','w') as f:\n",
    "        f.write('ensemble,min_precision,max_precision,mean_precision,std_precision,min_recall,max_recall,mean_recall,std_recall,min_accuracy,max_accuracy,mean_accuracy,std_accuracy')\n",
    "        f.write('\\n')\n",
    "    \n",
    "        f.write(f'ensemble,{min_prec},{max_prec},{mean_prec},{std_prec},{min_rec},{max_rec},{mean_rec},{std_rec},{min_acc},{max_acc},{mean_acc},{std_acc}')\n",
    "        f.write('\\n')\n",
    "        \n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a8758331",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 16.4 s, sys: 0 ns, total: 16.4 s\n",
      "Wall time: 1h 47min 6s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "ensemble_eval(train_enz_seq_file, None, label_file, train_feature_dirs, None, indhpoptfile, 'SVM', 5, False, Num_sim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c3384ac7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ensemble_store(enz_file, test_enz_file, label_file, train_feat_dirs, test_feat_dirs, hyper_param_file, base_algo, k, opt, N):\n",
    "    \"\"\"function stores the validation data in first line and prediction data in second line\"\"\"\n",
    "    pool = mp.Pool(mp.cpu_count())\n",
    "    \n",
    "    iter_func = zip([enz_file for _ in range(N)],\n",
    "           [test_enz_file for _ in range(N)],\n",
    "           [label_file for _ in range(N)],\n",
    "           [train_feat_dirs for _ in range(N)],\n",
    "           [test_feat_dirs for _ in range(N)],\n",
    "           [hyper_param_file for _ in range(N)],\n",
    "           [base_algo for _ in range(N)],\n",
    "           [k for _ in range(N)],\n",
    "           [opt for _ in range(N)],\n",
    "           range(N))\n",
    "\n",
    "    all_preds = list(pool.starmap(helper.ensemble_pred, iter_func))\n",
    "    \n",
    "    with open(\"../data/results/ensemble_preds.csv\",\"w\") as f:\n",
    "        for rs_pred in all_preds:\n",
    "            f.write(\",\".join(map(str,rs_pred[0])))\n",
    "            f.write(\"\\n\")\n",
    "            f.write(\",\".join(map(str,rs_pred[1])))\n",
    "            f.write(\"\\n\")\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0def8843",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.07 s, sys: 0 ns, total: 1.07 s\n",
      "Wall time: 1h 46min 28s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "ensemble_store(train_enz_seq_file, None, label_file, train_feature_dirs, None, indhpoptfile, 'SVM', 5, False, Num_sim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9161db45",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
